crates/analytics/src/opensearch.rs

```diff
@@ -0,0 +1,394 @@
+        match query_builder.query_type {
+            OpenSearchQuery::Msearch => {
+                let search_indexes = SearchIndex::iter();
+
+                let payload = query_builder
+                    .construct_payload(search_indexes.clone().collect())
+                    .change_context(OpenSearchError::ResponseError)?;
+
+                let mut payload_with_indexes: Vec<JsonBody<Value>> = vec![];
+                for (index_hit, index) in payload.iter().to_owned().zip(search_indexes) {
+                    payload_with_indexes.push(
+                        json!({"index": self.search_index_to_opensearch_index(index)}).into(),
+                    );
+                    payload_with_indexes.push(JsonBody::new(index_hit.clone()));
+                }
+
+                self.client
+                    .msearch(MsearchParts::None)
+                    .body(payload_with_indexes)
+                    .send()
+                    .await
+                    .change_context(OpenSearchError::ResponseError)
+            }
+            OpenSearchQuery::Search(index) => {
+                let payload = query_builder
+                    .clone()
+                    .construct_payload(vec![index])
+                    .change_context(OpenSearchError::ResponseError)?;
+
+                let final_payload = payload.first().unwrap_or(&Value::Null);
+
+                self.client
+                    .search(SearchParts::Index(&[
+                        &self.search_index_to_opensearch_index(index)
+                    ]))
+                    .from(query_builder.offset.unwrap_or(0))
+                    .size(query_builder.count.unwrap_or(10))
+                    .body(final_payload)
+                    .send()
+                    .await
+                    .change_context(OpenSearchError::ResponseError)
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl HealthCheck for OpenSearchClient {
+    async fn deep_health_check(&self) -> CustomResult<(), QueryExecutionError> {
+        let health = Cluster::new(&self.transport)
+            .health(ClusterHealthParts::None)
+            .send()
+            .await
+            .change_context(QueryExecutionError::DatabaseError)?
+            .json::<OpenSearchHealth>()
+            .await
+            .change_context(QueryExecutionError::DatabaseError)?;
+
+        if health.status != OpenSearchHealthStatus::Red {
+            Ok(())
+        } else {
+            Err(QueryExecutionError::DatabaseError.into())
+        }
+    }
+}
+
+impl OpenSearchIndexes {
+    pub fn validate(&self) -> Result<(), ApplicationError> {
+        use common_utils::ext_traits::ConfigExt;
+
+        common_utils::fp_utils::when(self.payment_attempts.is_default_or_empty(), || {
+            Err(ApplicationError::InvalidConfigurationValueError(
+                "Opensearch Payment Attempts index must not be empty".into(),
+            ))
+        })?;
```

Comment:
As a minor improvement, you can also import `when`:

```suggestion
        use common_utils::{ext_traits::ConfigExt, fp_utils::when};

        when(self.payment_attempts.is_default_or_empty(), || {
            Err(ApplicationError::InvalidConfigurationValueError(
                "Opensearch Payment Attempts index must not be empty".into(),
            ))
        })?;
```